File: python-compressor/examples/basic_example.py
"""
Basic ASON compression example
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from ason import SmartCompressor
import json


def main():
    # Create compressor
    compressor = SmartCompressor(indent=1, use_references=True)
    
    # Sample data
    data = {
        "customer": {
            "name": "John Doe",
            "email": "john.doe@example.com",
            "phone": "+1-555-0123",
            "tier": "gold"
        },
        "order": {
            "id": "ORD-12345",
            "date": "2024-01-15",
            "items": [
                {"product": "Widget A", "quantity": 2, "price": 29.99},
                {"product": "Widget B", "quantity": 1, "price": 49.99}
            ],
            "total": 109.97
        },
        "billing": {
            "email": "john.doe@example.com",
            "method": "credit_card"
        }
    }
    
    print("=" * 60)
    print("ASON Python Compressor - Basic Example")
    print("=" * 60)
    
    # Original JSON
    json_str = json.dumps(data, indent=2)
    print("\n📄 Original JSON:")
    print(json_str)
    print(f"\nSize: {len(json_str)} bytes")
    
    # Compress to ASON
    ason_str = compressor.compress(data)
    print("\n📦 Compressed ASON:")
    print(ason_str)
    print(f"\nSize: {len(ason_str)} bytes")
    
    # Compression stats
    result = compressor.compress_with_stats(data)
    print("\n📊 Compression Statistics:")
    print(f"  • Original tokens: {result['original_tokens']}")
    print(f"  • Compressed tokens: {result['compressed_tokens']}")
    print(f"  • Token reduction: {result['reduction_percent']}%")
    print(f"  • Size reduction: {result['stats']['size_reduction_percent']}%")
    
    # Round-trip validation
    decompressed = compressor.decompress(ason_str)
    print("\n✅ Round-trip validation:")
    print(f"  • Original matches decompressed: {data == decompressed}")
    
    # Optimization analysis
    opt_stats = compressor.get_optimization_stats(data)
    print("\n🔍 Optimization Analysis:")
    print(f"  • References detected: {opt_stats['references']['count']}")
    if opt_stats['references']['names']:
        print(f"    - {', '.join(opt_stats['references']['names'])}")
    
    print("\n" + "=" * 60)
    print("Example completed successfully!")
    print("=" * 60)


if __name__ == '__main__':
    main()

File: go-compressor/examples/basic_example.go
package main

import (
	"encoding/json"
	"fmt"
	
	"github.com/ason-format/ason/go-compressor/pkg/ason"
)

func main() {
	fmt.Println("=== ASON Go Compressor - Basic Example ===\n")
	
	// Create compressor with default options
	opts := ason.DefaultOptions()
	opts.Indent = 1
	compressor := ason.NewSmartCompressor(opts)
	
	// Example 1: Simple object
	fmt.Println("Example 1: Simple Object")
	fmt.Println("-------------------------")
	
	simpleData := map[string]interface{}{
		"name":  "Alice",
		"email": "alice@example.com",
		"age":   float64(30),
	}
	
	simpleJSON, _ := json.MarshalIndent(simpleData, "", "  ")
	fmt.Println("JSON:")
	fmt.Println(string(simpleJSON))
	
	simpleASON := compressor.Compress(simpleData)
	fmt.Println("\nASON:")
	fmt.Println(simpleASON)
	fmt.Println()
	
	// Example 2: Array of objects (tabular format)
	fmt.Println("Example 2: Tabular Array")
	fmt.Println("------------------------")
	
	tabularData := map[string]interface{}{
		"users": []interface{}{
			map[string]interface{}{"id": float64(1), "name": "Alice", "email": "alice@ex.com"},
			map[string]interface{}{"id": float64(2), "name": "Bob", "email": "bob@ex.com"},
			map[string]interface{}{"id": float64(3), "name": "Charlie", "email": "charlie@ex.com"},
		},
	}
	
	tabularJSON, _ := json.MarshalIndent(tabularData, "", "  ")
	fmt.Println("JSON:")
	fmt.Println(string(tabularJSON))
	
	tabularASON := compressor.Compress(tabularData)
	fmt.Println("\nASON:")
	fmt.Println(tabularASON)
	fmt.Println()
	
	// Example 3: Compression with stats
	fmt.Println("Example 3: Compression Stats")
	fmt.Println("----------------------------")
	
	statsData := map[string]interface{}{
		"products": []interface{}{
			map[string]interface{}{"id": float64(1), "name": "Laptop", "price": float64(999.99), "stock": float64(15)},
			map[string]interface{}{"id": float64(2), "name": "Mouse", "price": float64(29.99), "stock": float64(50)},
			map[string]interface{}{"id": float64(3), "name": "Keyboard", "price": float64(79.99), "stock": float64(30)},
			map[string]interface{}{"id": float64(4), "name": "Monitor", "price": float64(299.99), "stock": float64(20)},
		},
	}
	
	result := compressor.CompressWithStats(statsData)
	
	fmt.Println("ASON:")
	fmt.Println(result["ason"])
	
	stats := result["stats"].(map[string]interface{})
	fmt.Printf("\nOriginal tokens: %d\n", stats["original_tokens"])
	fmt.Printf("Compressed tokens: %d\n", stats["compressed_tokens"])
	fmt.Printf("Token reduction: %.2f%%\n", stats["reduction_percent"])
	fmt.Printf("Original size: %d bytes\n", stats["original_size"])
	fmt.Printf("Compressed size: %d bytes\n", stats["compressed_size"])
	fmt.Printf("Size reduction: %.2f%%\n", stats["size_reduction_pct"])
	fmt.Println()
	
	// Example 4: Round-trip validation
	fmt.Println("Example 4: Round-trip Validation")
	fmt.Println("--------------------------------")
	
	validationData := map[string]interface{}{
		"order": map[string]interface{}{
			"id":       "ORD-12345",
			"customer": "Alice Johnson",
			"items": []interface{}{
				map[string]interface{}{"product": "Widget", "quantity": float64(2), "price": float64(19.99)},
				map[string]interface{}{"product": "Gadget", "quantity": float64(1), "price": float64(49.99)},
			},
			"total": float64(89.97),
		},
	}
	
	validation := compressor.ValidateRoundTrip(validationData)
	
	if validation["valid"].(bool) {
		fmt.Println("✓ Round-trip validation: PASSED")
		fmt.Println("\nCompressed ASON:")
		fmt.Println(validation["compressed"])
	} else {
		fmt.Println("✗ Round-trip validation: FAILED")
		if err, ok := validation["error"]; ok {
			fmt.Printf("Error: %v\n", err)
		}
	}
	
	fmt.Println("\n=== Examples Complete ===")
}
